### PR TITLE
CI: Enable 32bit unit test (simplified)

### PR DESCRIPTION
**Pull request by falbrechtskirchinger**
Replaces #3529.

### CODE REVIEW

File: cmake/ci.cmake [hunk 1]
```diff
@@ -577,6 +584,13 @@ add_custom_target(ci_test_amalgamation
     COMMAND ${ASTYLE_TOOL} ${ASTYLE_FLAGS} ${INDENT_FILES}
     COMMAND cd ${PROJECT_SOURCE_DIR} && for FILE in `find . -name '*.orig'`\; do false \; done
 
+    COMMAND CXX=${GCC_TOOL} CXXFLAGS="${GCC_CXXFLAGS}" ${CMAKE_COMMAND}
```
Comment (nlohmann) [inline cmake/ci.cmake:587]: Why did you add this to `ci_test_amalgamation`?

File: tests/CMakeLists.txt [hunk 2]
```diff
@@ -124,34 +116,42 @@ message(STATUS "${msg}")
...
-    target_include_directories(test_main32 ${test_main_INCLUDE_DIRECTORIES})
-    target_link_libraries(test_main32 ${test_main_LINK_LIBRARIES})
-    target_link_options(test_main32 PUBLIC -m32)
-
-    json_test_add_test_for("src/unit-32bit.cpp" MAIN test_main32
-        CXX_STANDARDS ${test_cxx_standards} ${test_force})
+if(NOT "${json_32bit_test}" STREQUAL ONLY)
+    # test legacy comparison of discarded values
+    json_test_set_test_options(test-comparison_legacy
+        COMPILE_DEFINITIONS JSON_USE_LEGACY_DISCARDED_VALUE_COMPARISON=1
```
Comment (nlohmann) [inline tests/CMakeLists.txt:147]: Why does this depend on `json_32bit_test`?

File: cmake/ci.cmake (hunk 1 above)
Comment (falbrechtskirchinger) [reply cmake/ci.cmake:587]: It seemed related. Amalgamation should be the only reason why the single-header could possibly behave differently than the multi-header version.

File: tests/CMakeLists.txt (hunk 2 above)
Comment (falbrechtskirchinger) [reply tests/CMakeLists.txt:147]: It depends on `JSON_32bitTest=ONLY`. No other tests should be built. Technically, that also applies to the CMake tests below.

File: cmake/ci.cmake (hunk 1 above)
Comment (nlohmann) [reply cmake/ci.cmake:587]: But the test only aims at the structure. The goal is not to detect different behavior, but is to make sure changes are applied to both the `include` and the `single_include` folder.

File: tests/CMakeLists.txt (hunk 2 above)
Comment (nlohmann) [reply tests/CMakeLists.txt:147]: This feels over-engineered to me. Why do we need the flag in the first place?

Wouldn't it be sufficient to check, inside the coverage target, to add the 32 bit unit test with `-m32` flag?

File: cmake/ci.cmake (hunk 1 above)
Comment (falbrechtskirchinger) [reply cmake/ci.cmake:587]: I guess I just wanted to avoid adding another job to the workflow.

Look forward to an issue (and PR) on `json-ci` soon. Preview:
> The `json-ci` image is about 14GB in size. The time spent initializing containers is 170 seconds (average over ~350 jobs) or almost 3 minutes. That amounts to a ridiculous 2.5 hours per workflow run.

Comment (nlohmann) [reply cmake/ci.cmake:587]: Oh I so agree! We need to cut off the easy targets from the harder ones. I guess 2-3 Debian images could be sufficient to cover 90% of the compilers. Then we'd only need 1 image for the bleeding edge stuff and the specific tools.

File: tests/CMakeLists.txt (hunk 2 above)
Comment (falbrechtskirchinger) [reply tests/CMakeLists.txt:147]: It's not possible to "just" build a single unit test with `-m32` because that also needs the shared code (test_main) built with `-m32`. That's what I wanted to avoid in this "simplified" version. Instead, we *only* build the 32bit test. But I do have another idea. Just add a `return()` to the `CMakeLists.txt` and skip everything else in the file. I'll check if that's possible.

File: cmake/ci.cmake (hunk 1 above)
Comment (falbrechtskirchinger) [reply cmake/ci.cmake:587]: Do you want a test for the single-header at all?

Comment (nlohmann) [reply cmake/ci.cmake:587]: Good point. So far, we tested it by chance somewhere, because we did not switch on the multiple-header version in all of the tests. Now that we have the multiple-header version as default, a dedicated test for the amalgamated version should be in order.
